### ISSUE TITLE
Suggestion to migrate CI tests to GitHub 

### ISSUE THREAD

**Issue by ikappaki** [clojure-emacs/cider#3207]
Hi,

I believe there is some merit migrating the CI tests to a GitHub workflow, is there any appetite for it?

1. Easy access via the `Actions` menu in GitHub.
2. Currently macos tests are running on GitHub while the rest run on CircleCi, we can get test results under a single badge. 
3. GitHub appears to give unlimited credit for running jobs, thus we can extend the test coverage to include all supported versions across all architectures.

I have created a simple demo workflow to showcase the outcome https://github.com/ikappaki/cider/blob/a4de8824d8e3cdbc8756660fab78f883eea60bfd/.github/workflows/test.yml

It is configured to test what I believe are CIDER supported Emacs versions across macos, ubuntu and windows. There is also Emacs master setup only for ubuntu. 

![image](https://user-images.githubusercontent.com/34983288/169642740-381fcfcf-842c-474d-b7fb-f38c77541369.png)

Thanks

**Comment by bbatsov** [comment]
I'm totally fine with the idea of moving to GHA. (mostly because I don't like mixing CI tools across my projects and in general I've been slowly moving in the direction of GHA)

**Comment by vemv** [comment]
A tdlr would be to look at the big picture - what do we want for clojure-emacs (which most likely benefit from homogeneity), what do we have now, who is actively maintaining their CIs?

**Comment by ikappaki** [comment]
> A tdlr would be to look at the big picture - what do we want for clojure-emacs (which most likely benefit from homogeneity), what do we have now, who is actively maintaining their CIs?
> 
> From that point of view, I wouldn't want to introduce problems while our current situation is non-problematic. Churn is expensive, and our time is limited. I'd love to spend the little "OSS time" I have on actual feature development.

HI @vemv, no worries, this was only a suggestion anyway that I brought up for discussion. I thought of it as an improvement over the current CIDER CircleCI setup -- not that CCI is problematic, but GHA feels better IMHO. I've created a GHA setup with the last three version (26.3, 27.2 and 28.1) running all tests/linters/byte compilers accross macos, ubuntu, win and a snapshot run for Ubuntu. I will keep the branch up (that I was just about to PR) for a while in case it comes up at a later while considering the `bigger picture` or there is a change of sentiment: https://github.com/ikappaki/cider/tree/tech/github-ci-all

Thanks

**Comment by ikappaki** [comment]
Here is what appears to be another limitation/restriction with CCI. I can't see the CircleCI results on the PR, and I assume this is the same for the reviewers? i..e the reviewer does not know whether the changes is passing CI tests? It only shows the results of the GitHub Actions:

![image](https://user-images.githubusercontent.com/34983288/170816184-79cd00e3-a081-438c-992a-7a08b743ef93.png)

**Comment by vemv** [comment]
I think CCI does not run if you already have a forked project running in the same CCI account

i.e. CCI won't build a project twice (one for your fork, another for upstream).

Generally this isn't an issue as most people won't set up a CCI project for their forks.

**Comment by vemv** [comment]
> I'd like us to focus here on more technical merits and less philosophical ones

We could even start by wondering if there's a problem at all to begin with. If not we can compare all day long but not be solving a problem.

The only practical problem I perceive is that checks are not enforced for merging, so CI has a limited role in preventing mistakes. At the same time, that topic has been discussed time and again. Personally I'm quite fine with things as they are.

**Comment by ikappaki** [comment]
> I think CCI does not run if you already have a forked project running in your personal CCI account
> 
> i.e. CCI won't build a project twice (one for your fork, another for upstream).
> 
> Generally this isn't an issue as most people won't set up a CCI project for their forks.
> 
> Edit: GHA is not too different - it requires explicit build activation for forks, because it also doesn't want to run things too much

Hi @vemv, just for me to understand the behavior, if I haven't had a CCI account (or I didn't enable it somehow for a particular project), the CIDER PR would have run the tests on CCI and thus the results would be visible to reviewer? Is this what's happening on the other open PR (number 3201) and I as such I can see the CCI results on the PR? I was puzzled by this ...

![image](https://user-images.githubusercontent.com/34983288/170816873-083f7d81-100a-4bde-920a-5fab13c7fd55.png)

Thanks

**Comment by vemv** [comment]
Yes, in order for builds to show up you have to disable your project fork in your CCI account

(having a CCI account of course doesn't disable builds per se)
